File: ch.hsr.ifs.testframework/src/ch/hsr/ifs/testframework/model/ModellBuilder.java
/*******************************************************************************
 * Copyright (c) 2007-2011, IFS Institute for Software, HSR Rapperswil,
 * Switzerland, http://ifs.hsr.ch
 * 
 * Permission to use, copy, and/or distribute this software for any
 * purpose without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 ******************************************************************************/
package ch.hsr.ifs.testframework.model;

import java.net.URI;
import java.net.URISyntaxException;

import org.eclipse.core.resources.IFile;
import org.eclipse.core.resources.IResource;
import org.eclipse.core.resources.ResourcesPlugin;
import org.eclipse.core.runtime.IPath;
import org.eclipse.debug.core.ILaunch;
import org.eclipse.jface.text.IRegion;

import ch.hsr.ifs.testframework.TestFrameworkPlugin;
import ch.hsr.ifs.testframework.event.TestEventHandler;

/**
 * @author egraf
 *
 */
public class ModellBuilder extends TestEventHandler {

	private final Model model = TestFrameworkPlugin.getModel();
	private final IPath rtPath;
	private TestCase lastTestCase;
	private TestCase currentTestCase;
	private final ILaunch launch;

	public ModellBuilder(IPath exePath, ILaunch launch) {
		super();
		this.rtPath = exePath;
		this.launch = launch;
	}
	public ModellBuilder(IPath path) {
		this(path, null);
	}

	@Override
	public void handleError(IRegion reg, String testName, String msg) {
		if(currentTestCase != null) {
			model.endCurrentTestCase(null, -1, msg, TestStatus.error, currentTestCase);
			endTestCase();
		}else {
			unexpectedTestCaseEnd();
		}
	}
	private void unexpectedTestCaseEnd() {
		if(lastTestCase != null) {
			model.endCurrentTestCase(null, -1, Messages.ModellBuilder_0, TestStatus.error, lastTestCase);
		}

	}
	private void endTestCase() {
		lastTestCase = currentTestCase;
		currentTestCase = null;
	}

	@Override
	public void handleSuccess(IRegion reg, String name, String msg) {
		if(currentTestCase != null) {
			model.endCurrentTestCase(null, -1, msg, TestStatus.success, currentTestCase);
			endTestCase();
		}else {
			unexpectedTestCaseEnd();
		}
	}

	@Override
	public void handleEnding(IRegion reg, String suitename) {
		model.endSuite();
	}

	@Override
	public void handleBeginning(IRegion reg, String suitename, String suitesize) {
		model.startSuite(new TestSuite(suitename, Integer.parseInt(suitesize), TestStatus.running));
	}

	@Override
	public void handleFailure(IRegion reg, String testName, String fileName, String lineNo, String reason){
		if(currentTestCase != null) {

			IPath filePath = getWorkspaceFile(fileName, rtPath);		
			
			IFile file = ResourcesPlugin.getWorkspace().getRoot().getFileForLocation(filePath);
			if(file == null){
				try {
					IFile[] files = ResourcesPlugin.getWorkspace().getRoot().findFilesForLocationURI(new URI("file:" + filePath.toPortableString()), IResource.FILE);
					if(files.length > 0)
					{
						file = files[0];
					}
				} catch (URISyntaxException e) {
				}
			}
			int lineNumber = Integer.parseInt(lineNo);
			model.endCurrentTestCase(file, lineNumber, reason, TestStatus.failure, currentTestCase);
			endTestCase();
		} else {
			unexpectedTestCaseEnd();
		}

	}

	@Override
	public void handleTestStart(IRegion reg, String suitename) {
		currentTestCase = new TestCase(suitename);
		model.addTest(currentTestCase);
	}

	@Override
	public void handleSessionEnd() {
		model.endSession(currentTestCase);
	}

	@Override
	public void handleSessionStart() {
		model.startSession(launch);
	}

}

File: CUTEProject/bundles/ch.hsr.ifs.testframework/src/ch/hsr/ifs/testframework/model/ModellBuilder.kt
/*******************************************************************************
 * Copyright (c) 2007-2011, IFS Institute for Software, HSR Rapperswil,
 * Switzerland, http://ifs.hsr.ch
 *
 * Permission to use, copy, and/or distribute this software for any
 * purpose without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 ******************************************************************************/
package ch.hsr.ifs.testframework.model

import java.net.URI
import java.net.URISyntaxException

import org.eclipse.core.resources.IFile
import org.eclipse.core.resources.IResource
import org.eclipse.core.resources.IWorkspaceRoot
import org.eclipse.core.resources.ResourcesPlugin
import org.eclipse.core.runtime.IPath
import org.eclipse.debug.core.ILaunch
import org.eclipse.jface.text.IRegion

import ch.hsr.ifs.testframework.TestFrameworkPlugin
import ch.hsr.ifs.testframework.event.TestEventHandler
import ch.hsr.ifs.testframework.model.Model

/**
 * @author egraf
 *
 */
class ModellBuilder @JvmOverloads constructor(private val exePath: IPath, private val launch: ILaunch? = null) : TestEventHandler() {
 
   private val model = TestFrameworkPlugin.getModel()
   private var lastTestCase: TestCase? = null
   private var currentTestCase: TestCase? = null

   override fun handleError(reg: IRegion, testName: String , msg: String) {
      when(currentTestCase) {
         null -> unexpectedTestCaseEnd()
         else -> {
            model.endCurrentTestCase(null, -1, msg, TestStatus.error, currentTestCase!!)
            endTestCase()
         }
      }
   }

   fun unexpectedTestCaseEnd() {
      lastTestCase?.let{model.endCurrentTestCase(null, -1, Messages.ModellBuilder_0, TestStatus.error, it)}
   }

   fun endTestCase() {
      lastTestCase = currentTestCase
      currentTestCase = null
   }

   override fun handleSuccess(reg: IRegion, name: String, msg: String) {
      when(currentTestCase) {
         null -> unexpectedTestCaseEnd()
         else -> {
            model.endCurrentTestCase(null, -1, msg, TestStatus.success, currentTestCase!!)
            endTestCase()
         }
      }
   }

   override fun handleEnding(reg: IRegion, suitename: String) {
      model.endSuite()
   }

   override fun handleBeginning(reg: IRegion , suitename: String, suitesize: String) {
      model.startSuite(TestSuite(suitename, Integer.parseInt(suitesize), TestStatus.running))
   }


   override fun handleFailure(reg:IRegion , testName: String, fileName: String, lineNo: String, reason: String) {
      when(currentTestCase) {
         null -> unexpectedTestCaseEnd()
         else -> {
            val filePath = getWorkspaceFile(fileName, exePath)
   
            val root = ResourcesPlugin.getWorkspace().getRoot()
            var file = root.getFileForLocation(filePath)
            if (file == null) {
               try {
                  val files = root.findFilesForLocationURI(URI("file:" + filePath.toPortableString()), IResource.FILE)
                  if (files.size > 0) {
                     file = files[0]
                  }
               } catch (e: URISyntaxException) {}
            }
            val lineNumber = Integer.parseInt(lineNo)
            model.endCurrentTestCase(file, lineNumber, reason, TestStatus.failure, currentTestCase!!)
            endTestCase()
         }
      }
   }

   override fun handleTestStart(reg: IRegion, testName: String) {
      currentTestCase = TestCase(testName)
      model.addTest(currentTestCase!!)
   }

   override fun handleSessionEnd() = model.endSession(currentTestCase)

   override fun handleSessionStart() = model.startSession(launch)

}
